ICE "no index for a field"

Found with a [modified fuzz-rustc](https://github.com/jruderman/fuzz-rustc)

### Code

```Rust
struct S {
    a: u32,
}

fn main() {
    let s1 = S { a: 1 };

    let _ = || {
        let s2 = Oops { a: 2, ..s1 };
    };
}
```

### Error output

```
error[E0422]: cannot find struct, variant or union type `Oops` in this scope
 --> src/main.rs:9:18
  |
9 |         let s2 = Oops { a: 2, ..s1 };
  |                  ^^^^ not found in this scope

thread '<unnamed>' panicked at 'no index for a field', compiler/rustc_middle/src/ty/mod.rs:2326:9
```

<details><summary><strong>Backtrace</strong></summary>
<p>

```
stack backtrace:
   0:        0x105fd1c52 - <std::sys_common::backtrace::_print::DisplayBacktrace as core::fmt::Display>::fmt::h65371da71ea634a1
   1:        0x1060306aa - core::fmt::write::h42d20afb3c5ee095
   2:        0x105fc3f0c - std::io::Write::write_fmt::h8d5b26d0e53ddd7f
   3:        0x105fd1a1a - std::sys_common::backtrace::print::hb5fe5178795d9f17
   4:        0x105fd4d46 - std::panicking::default_hook::{{closure}}::hff2fd60e97cc3869
   5:        0x105fd4a97 - std::panicking::default_hook::hc3f9a1de84e0ed9d
   6:        0x1148f4aed - rustc_driver[13343b5551cdacdb]::DEFAULT_HOOK::{closure#0}::{closure#0}
   7:        0x105fd552d - std::panicking::rust_panic_with_hook::h1c5a191a8000993f
   8:        0x105fd52d3 - std::panicking::begin_panic_handler::{{closure}}::hd5b348ff58d0d76f
   9:        0x105fd20e8 - std::sys_common::backtrace::__rust_end_short_backtrace::h724f40d3ad4af19e
  10:        0x105fd4f9d - _rust_begin_unwind
  11:        0x10605d443 - core::panicking::panic_fmt::hc08710230bc1c166
  12:        0x10602d1db - core::panicking::panic_display::h484762008229854f
  13:        0x10602d18c - core::panicking::panic_str::h21235dd98dd92f90
  14:        0x10605d409 - core::option::expect_failed::hfda067ca1d1574cb
  15:        0x11910855a - <rustc_middle[fb39214eb358b7b2]::ty::context::TyCtxt>::field_index
  16:        0x117234dfc - <rustc_hir_typeck[280d4fd6d44492fc]::expr_use_visitor::ExprUseVisitor>::walk_expr
  17:        0x1172357f5 - <rustc_hir_typeck[280d4fd6d44492fc]::expr_use_visitor::ExprUseVisitor>::walk_local::<<rustc_hir_typeck[280d4fd6d44492fc]::expr_use_visitor::ExprUseVisitor>::walk_stmt::{closure#0}>
  18:        0x1172353e6 - <rustc_hir_typeck[280d4fd6d44492fc]::expr_use_visitor::ExprUseVisitor>::walk_expr
  19:        0x11723310c - <rustc_hir_typeck[280d4fd6d44492fc]::expr_use_visitor::ExprUseVisitor>::consume_expr
  20:        0x1172362d7 - <rustc_hir_typeck[280d4fd6d44492fc]::expr_use_visitor::ExprUseVisitor>::consume_body
  21:        0x117112ff6 - <rustc_hir_typeck[280d4fd6d44492fc]::fn_ctxt::FnCtxt>::analyze_closure
  22:        0x11721c968 - <rustc_hir_typeck[280d4fd6d44492fc]::upvar::InferBorrowKindVisitor as rustc_hir[30bdd25a7791883c]::intravisit::Visitor>::visit_expr
  23:        0x11714c7c8 - rustc_hir[30bdd25a7791883c]::intravisit::walk_local::<rustc_hir_typeck[280d4fd6d44492fc]::upvar::InferBorrowKindVisitor>
  24:        0x117154e4c - rustc_hir[30bdd25a7791883c]::intravisit::walk_expr::<rustc_hir_typeck[280d4fd6d44492fc]::upvar::InferBorrowKindVisitor>
  25:        0x1171e89ea - <rustc_hir_typeck[280d4fd6d44492fc]::inherited::InheritedBuilder>::enter::<rustc_hir_typeck[280d4fd6d44492fc]::typeck_with_fallback<rustc_hir_typeck[280d4fd6d44492fc]::typeck::{closure#0}>::{closure#0}::{closure#1}, &rustc_middle[fb39214eb358b7b2]::ty::context::TypeckResults>
  26:        0x11717ec4a - rustc_hir_typeck[280d4fd6d44492fc]::typeck
  27:        0x1180d3669 - rustc_query_system[1f9dfa1b3b6feb8f]::query::plumbing::try_execute_query::<rustc_query_impl[bea551e74a118b14]::plumbing::QueryCtxt, rustc_query_system[1f9dfa1b3b6feb8f]::query::caches::DefaultCache<rustc_span[3cd6c9c9c366901b]::def_id::LocalDefId, &rustc_middle[fb39214eb358b7b2]::ty::context::TypeckResults>>
  28:        0x1181cfb0c - rustc_query_system[1f9dfa1b3b6feb8f]::query::plumbing::get_query::<rustc_query_impl[bea551e74a118b14]::queries::typeck, rustc_query_impl[bea551e74a118b14]::plumbing::QueryCtxt>
  29:        0x11712ffcb - rustc_data_structures[65115343bf925642]::sync::par_for_each_in::<&[rustc_span[3cd6c9c9c366901b]::def_id::LocalDefId], <rustc_middle[fb39214eb358b7b2]::hir::map::Map>::par_body_owners<rustc_hir_typeck[280d4fd6d44492fc]::typeck_item_bodies::{closure#0}>::{closure#0}>
  30:        0x11717e3cd - rustc_hir_typeck[280d4fd6d44492fc]::typeck_item_bodies
  31:        0x118163e59 - rustc_query_system[1f9dfa1b3b6feb8f]::query::plumbing::try_execute_query::<rustc_query_impl[bea551e74a118b14]::plumbing::QueryCtxt, rustc_query_system[1f9dfa1b3b6feb8f]::query::caches::DefaultCache<(), ()>>
  32:        0x1181c43d9 - rustc_query_system[1f9dfa1b3b6feb8f]::query::plumbing::get_query::<rustc_query_impl[bea551e74a118b14]::queries::typeck_item_bodies, rustc_query_impl[bea551e74a118b14]::plumbing::QueryCtxt>
  33:        0x1173045ae - <rustc_session[d3dc83f7882946da]::session::Session>::time::<(), rustc_hir_analysis[65e5df12102afc97]::check_crate::{closure#7}>
  34:        0x11740de8f - rustc_hir_analysis[65e5df12102afc97]::check_crate
  35:        0x1149b155a - rustc_interface[bc80ce5eea40f24e]::passes::analysis
  36:        0x118155c1c - rustc_query_system[1f9dfa1b3b6feb8f]::query::plumbing::try_execute_query::<rustc_query_impl[bea551e74a118b14]::plumbing::QueryCtxt, rustc_query_system[1f9dfa1b3b6feb8f]::query::caches::DefaultCache<(), core[5f16df7105d54e]::result::Result<(), rustc_errors[56d5964ce0535c75]::ErrorGuaranteed>>>
  37:        0x1181cfeb9 - rustc_query_system[1f9dfa1b3b6feb8f]::query::plumbing::get_query::<rustc_query_impl[bea551e74a118b14]::queries::analysis, rustc_query_impl[bea551e74a118b14]::plumbing::QueryCtxt>
  38:        0x1148b9a72 - <rustc_interface[bc80ce5eea40f24e]::passes::QueryContext>::enter::<rustc_driver[13343b5551cdacdb]::run_compiler::{closure#1}::{closure#2}::{closure#3}, core[5f16df7105d54e]::result::Result<(), rustc_errors[56d5964ce0535c75]::ErrorGuaranteed>>
  39:        0x1148cb94d - rustc_span[3cd6c9c9c366901b]::with_source_map::<core[5f16df7105d54e]::result::Result<(), rustc_errors[56d5964ce0535c75]::ErrorGuaranteed>, rustc_interface[bc80ce5eea40f24e]::interface::run_compiler<core[5f16df7105d54e]::result::Result<(), rustc_errors[56d5964ce0535c75]::ErrorGuaranteed>, rustc_driver[13343b5551cdacdb]::run_compiler::{closure#1}>::{closure#0}::{closure#1}>
  40:        0x1148c222c - <scoped_tls[9f05bfe7696cff9b]::ScopedKey<rustc_span[3cd6c9c9c366901b]::SessionGlobals>>::set::<rustc_interface[bc80ce5eea40f24e]::interface::run_compiler<core[5f16df7105d54e]::result::Result<(), rustc_errors[56d5964ce0535c75]::ErrorGuaranteed>, rustc_driver[13343b5551cdacdb]::run_compiler::{closure#1}>::{closure#0}, core[5f16df7105d54e]::result::Result<(), rustc_errors[56d5964ce0535c75]::ErrorGuaranteed>>
  41:        0x11488a19a - std[b8dd7466c60cfdf8]::sys_common::backtrace::__rust_begin_short_backtrace::<rustc_interface[bc80ce5eea40f24e]::util::run_in_thread_pool_with_globals<rustc_interface[bc80ce5eea40f24e]::interface::run_compiler<core[5f16df7105d54e]::result::Result<(), rustc_errors[56d5964ce0535c75]::ErrorGuaranteed>, rustc_driver[13343b5551cdacdb]::run_compiler::{closure#1}>::{closure#0}, core[5f16df7105d54e]::result::Result<(), rustc_errors[56d5964ce0535c75]::ErrorGuaranteed>>::{closure#0}::{closure#0}, core[5f16df7105d54e]::result::Result<(), rustc_errors[56d5964ce0535c75]::ErrorGuaranteed>>
  42:        0x11486c96b - <<std[b8dd7466c60cfdf8]::thread::Builder>::spawn_unchecked_<rustc_interface[bc80ce5eea40f24e]::util::run_in_thread_pool_with_globals<rustc_interface[bc80ce5eea40f24e]::interface::run_compiler<core[5f16df7105d54e]::result::Result<(), rustc_errors[56d5964ce0535c75]::ErrorGuaranteed>, rustc_driver[13343b5551cdacdb]::run_compiler::{closure#1}>::{closure#0}, core[5f16df7105d54e]::result::Result<(), rustc_errors[56d5964ce0535c75]::ErrorGuaranteed>>::{closure#0}::{closure#0}, core[5f16df7105d54e]::result::Result<(), rustc_errors[56d5964ce0535c75]::ErrorGuaranteed>>::{closure#1} as core[5f16df7105d54e]::ops::function::FnOnce<()>>::call_once::{shim:vtable#0}
  43:        0x105fde8e7 - std::sys::unix::thread::Thread::new::thread_start::h459496ec4dc9999f
  44:     0x7ff80d84d4e1 - __pthread_start
```

</p>
</details>

### Regression

Regression in nightly-2022-11-06, somewhere in rollup commit 452cf4f7109f58433ac38be7d3da527408571054

### Version

```
rustc 1.67.0-nightly (b833ad56f 2022-11-18)
binary: rustc
commit-hash: b833ad56f46a0bbe0e8729512812a161e7dae28a
commit-date: 2022-11-18
host: x86_64-apple-darwin
release: 1.67.0-nightly
LLVM version: 15.0.4
```

--- a/compiler/rustc_hir_typeck/src/expr.rs
+++ b/compiler/rustc_hir_typeck/src/expr.rs
@@ -1646,6 +1646,9 @@ impl<'a, 'tcx> FnCtxt<'a, 'tcx> {
         // the fields with the base_expr. This could cause us to hit errors later
         // when certain fields are assumed to exist that in fact do not.
         if error_happened {
+            if let Some(base_expr) = base_expr {
+                self.check_expr(base_expr);
+            }
             return;
         }
 

--- a/compiler/rustc_hir_typeck/src/expr_use_visitor.rs
+++ b/compiler/rustc_hir_typeck/src/expr_use_visitor.rs
@@ -523,6 +523,11 @@ impl<'a, 'tcx> ExprUseVisitor<'a, 'tcx> {
         // Consume the expressions supplying values for each field.
         for field in fields {
             self.consume_expr(field.expr);
+
+            // The struct path probably didn't resolve
+            if self.mc.typeck_results.opt_field_index(field.hir_id).is_none() {
+                self.tcx().sess.delay_span_bug(field.span, "couldn't resolve index for field");
+            }
         }
 
         let with_expr = match *opt_with {
@@ -542,7 +547,7 @@ impl<'a, 'tcx> ExprUseVisitor<'a, 'tcx> {
                 for (f_index, with_field) in adt.non_enum_variant().fields.iter().enumerate() {
                     let is_mentioned = fields
                         .iter()
-                        .any(|f| self.mc.typeck_results.field_index(f.hir_id) == f_index);
+                        .any(|f| self.mc.typeck_results.opt_field_index(f.hir_id) == Some(f_index));
                     if !is_mentioned {
                         let field_place = self.mc.cat_projection(
                             &*with_expr,

--- a/compiler/rustc_hir_typeck/src/expr_use_visitor.rs
+++ b/compiler/rustc_hir_typeck/src/expr_use_visitor.rs
@@ -540,9 +540,9 @@ impl<'a, 'tcx> ExprUseVisitor<'a, 'tcx> {
             ty::Adt(adt, substs) if adt.is_struct() => {
                 // Consume those fields of the with expression that are needed.
                 for (f_index, with_field) in adt.non_enum_variant().fields.iter().enumerate() {
-                    let is_mentioned = fields.iter().any(|f| {
-                        self.tcx().field_index(f.hir_id, self.mc.typeck_results) == f_index
-                    });
+                    let is_mentioned = fields
+                        .iter()
+                        .any(|f| self.mc.typeck_results.field_index(f.hir_id) == f_index);
                     if !is_mentioned {
                         let field_place = self.mc.cat_projection(
                             &*with_expr,

--- a/compiler/rustc_lint/src/builtin.rs
+++ b/compiler/rustc_lint/src/builtin.rs
@@ -259,7 +259,7 @@ impl<'tcx> LateLintPass<'tcx> for NonShorthandFieldPatterns {
                 }
                 if let PatKind::Binding(binding_annot, _, ident, None) = fieldpat.pat.kind {
                     if cx.tcx.find_field_index(ident, &variant)
-                        == Some(cx.tcx.field_index(fieldpat.hir_id, cx.typeck_results()))
+                        == Some(cx.typeck_results().field_index(fieldpat.hir_id))
                     {
                         cx.struct_span_lint(
                             NON_SHORTHAND_FIELD_PATTERNS,

--- a/compiler/rustc_middle/src/ty/context.rs
+++ b/compiler/rustc_middle/src/ty/context.rs
@@ -671,6 +671,14 @@ impl<'tcx> TypeckResults<'tcx> {
         LocalTableInContextMut { hir_owner: self.hir_owner, data: &mut self.field_indices }
     }
 
+    pub fn field_index(&self, id: hir::HirId) -> usize {
+        self.field_indices().get(id).cloned().expect("no index for a field")
+    }
+
+    pub fn opt_field_index(&self, id: hir::HirId) -> Option<usize> {
+        self.field_indices().get(id).cloned()
+    }
+
     pub fn user_provided_types(&self) -> LocalTableInContext<'_, CanonicalUserType<'tcx>> {
         LocalTableInContext { hir_owner: self.hir_owner, data: &self.user_provided_types }
     }

--- a/compiler/rustc_middle/src/ty/mod.rs
+++ b/compiler/rustc_middle/src/ty/mod.rs
@@ -2228,10 +2228,6 @@ impl<'tcx> TyCtxt<'tcx> {
         }
     }
 
-    pub fn field_index(self, hir_id: hir::HirId, typeck_results: &TypeckResults<'_>) -> usize {
-        typeck_results.field_indices().get(hir_id).cloned().expect("no index for a field")
-    }
-
     pub fn find_field_index(self, ident: Ident, variant: &VariantDef) -> Option<usize> {
         variant
             .fields

--- a/compiler/rustc_mir_build/src/thir/cx/expr.rs
+++ b/compiler/rustc_mir_build/src/thir/cx/expr.rs
@@ -704,7 +704,7 @@ impl<'tcx> Cx<'tcx> {
             hir::ExprKind::Field(ref source, ..) => ExprKind::Field {
                 lhs: self.mirror_expr(source),
                 variant_index: VariantIdx::new(0),
-                name: Field::new(tcx.field_index(expr.hir_id, self.typeck_results)),
+                name: Field::new(self.typeck_results.field_index(expr.hir_id)),
             },
             hir::ExprKind::Cast(ref source, ref cast_ty) => {
                 // Check for a user-given type annotation on this `cast`
@@ -1079,7 +1079,7 @@ impl<'tcx> Cx<'tcx> {
         fields
             .iter()
             .map(|field| FieldExpr {
-                name: Field::new(self.tcx.field_index(field.hir_id, self.typeck_results)),
+                name: Field::new(self.typeck_results.field_index(field.hir_id)),
                 expr: self.mirror_expr(field.expr),
             })
             .collect()

--- a/compiler/rustc_mir_build/src/thir/pattern/mod.rs
+++ b/compiler/rustc_mir_build/src/thir/pattern/mod.rs
@@ -321,7 +321,7 @@ impl<'a, 'tcx> PatCtxt<'a, 'tcx> {
                 let subpatterns = fields
                     .iter()
                     .map(|field| FieldPat {
-                        field: Field::new(self.tcx.field_index(field.hir_id, self.typeck_results)),
+                        field: Field::new(self.typeck_results.field_index(field.hir_id)),
                         pattern: self.lower_pattern(&field.pat),
                     })
                     .collect();

--- a/compiler/rustc_passes/src/dead.rs
+++ b/compiler/rustc_passes/src/dead.rs
@@ -124,7 +124,7 @@ impl<'tcx> MarkSymbolVisitor<'tcx> {
     fn handle_field_access(&mut self, lhs: &hir::Expr<'_>, hir_id: hir::HirId) {
         match self.typeck_results().expr_ty_adjusted(lhs).kind() {
             ty::Adt(def, _) => {
-                let index = self.tcx.field_index(hir_id, self.typeck_results());
+                let index = self.typeck_results().field_index(hir_id);
                 self.insert_def_id(def.non_enum_variant().fields[index].did);
             }
             ty::Tuple(..) => {}
@@ -208,7 +208,7 @@ impl<'tcx> MarkSymbolVisitor<'tcx> {
             if let PatKind::Wild = pat.pat.kind {
                 continue;
             }
-            let index = self.tcx.field_index(pat.hir_id, self.typeck_results());
+            let index = self.typeck_results().field_index(pat.hir_id);
             self.insert_def_id(variant.fields[index].did);
         }
     }
@@ -341,7 +341,7 @@ impl<'tcx> MarkSymbolVisitor<'tcx> {
     fn mark_as_used_if_union(&mut self, adt: ty::AdtDef<'tcx>, fields: &[hir::ExprField<'_>]) {
         if adt.is_union() && adt.non_enum_variant().fields.len() > 1 && adt.did().is_local() {
             for field in fields {
-                let index = self.tcx.field_index(field.hir_id, self.typeck_results());
+                let index = self.typeck_results().field_index(field.hir_id);
                 self.insert_def_id(adt.non_enum_variant().fields[index].did);
             }
         }

--- a/compiler/rustc_privacy/src/lib.rs
+++ b/compiler/rustc_privacy/src/lib.rs
@@ -1065,9 +1065,9 @@ impl<'tcx> Visitor<'tcx> for NamePrivacyVisitor<'tcx> {
                 // are checked for privacy (RFC 736). Rather than computing the set of
                 // unmentioned fields, just check them all.
                 for (vf_index, variant_field) in variant.fields.iter().enumerate() {
-                    let field = fields.iter().find(|f| {
-                        self.tcx.field_index(f.hir_id, self.typeck_results()) == vf_index
-                    });
+                    let field = fields
+                        .iter()
+                        .find(|f| self.typeck_results().field_index(f.hir_id) == vf_index);
                     let (use_ctxt, span) = match field {
                         Some(field) => (field.ident.span, field.span),
                         None => (base.span, base.span),
@@ -1077,7 +1077,7 @@ impl<'tcx> Visitor<'tcx> for NamePrivacyVisitor<'tcx> {
             } else {
                 for field in fields {
                     let use_ctxt = field.ident.span;
-                    let index = self.tcx.field_index(field.hir_id, self.typeck_results());
+                    let index = self.typeck_results().field_index(field.hir_id);
                     self.check_field(use_ctxt, field.span, adt, &variant.fields[index], false);
                 }
             }
@@ -1093,7 +1093,7 @@ impl<'tcx> Visitor<'tcx> for NamePrivacyVisitor<'tcx> {
             let variant = adt.variant_of_res(res);
             for field in fields {
                 let use_ctxt = field.ident.span;
-                let index = self.tcx.field_index(field.hir_id, self.typeck_results());
+                let index = self.typeck_results().field_index(field.hir_id);
                 self.check_field(use_ctxt, field.span, adt, &variant.fields[index], false);
             }
         }

--- a/src/test/ui/async-await/drop-track-bad-field-in-fru.rs
+++ b/src/test/ui/async-await/drop-track-bad-field-in-fru.rs
@@ -1,0 +1,10 @@
+// compile-flags: -Zdrop-tracking
+// edition: 2021
+
+fn main() {}
+
+async fn foo() {
+    None { value: (), ..Default::default() }.await;
+    //~^ ERROR `Option<_>` is not a future
+    //~| ERROR variant `Option<_>::None` has no field named `value`
+}

--- a/src/test/ui/async-await/drop-track-bad-field-in-fru.stderr
+++ b/src/test/ui/async-await/drop-track-bad-field-in-fru.stderr
@@ -1,0 +1,23 @@
+error[E0559]: variant `Option<_>::None` has no field named `value`
+  --> $DIR/drop-track-bad-field-in-fru.rs:7:12
+   |
+LL |     None { value: (), ..Default::default() }.await;
+   |            ^^^^^ `Option<_>::None` does not have this field
+
+error[E0277]: `Option<_>` is not a future
+  --> $DIR/drop-track-bad-field-in-fru.rs:7:45
+   |
+LL |     None { value: (), ..Default::default() }.await;
+   |                                             ^^^^^^
+   |                                             |
+   |                                             `Option<_>` is not a future
+   |                                             help: remove the `.await`
+   |
+   = help: the trait `Future` is not implemented for `Option<_>`
+   = note: Option<_> must be a future or must implement `IntoFuture` to be awaited
+   = note: required for `Option<_>` to implement `IntoFuture`
+
+error: aborting due to 2 previous errors
+
+Some errors have detailed explanations: E0277, E0559.
+For more information about an error, try `rustc --explain E0277`.

--- a/src/test/ui/structs/unresolved-struct-with-fru.rs
+++ b/src/test/ui/structs/unresolved-struct-with-fru.rs
@@ -1,0 +1,12 @@
+struct S {
+    a: u32,
+}
+
+fn main() {
+    let s1 = S { a: 1 };
+
+    let _ = || {
+        let s2 = Oops { a: 2, ..s1 };
+        //~^ ERROR cannot find struct, variant or union type `Oops` in this scope
+    };
+}

--- a/src/test/ui/structs/unresolved-struct-with-fru.stderr
+++ b/src/test/ui/structs/unresolved-struct-with-fru.stderr
@@ -1,0 +1,9 @@
+error[E0422]: cannot find struct, variant or union type `Oops` in this scope
+  --> $DIR/unresolved-struct-with-fru.rs:9:18
+   |
+LL |         let s2 = Oops { a: 2, ..s1 };
+   |                  ^^^^ not found in this scope
+
+error: aborting due to previous error
+
+For more information about this error, try `rustc --explain E0422`.

